### PR TITLE
Refactor CreateDistributedTable()

### PR DESCRIPTION
**Pull request by onurctirtir**
Split the main logic that allows creating a Citus table into the
internal function CreateCitusTable().

Old CreateDistributedTable() function was assuming that it's creating
a reference table when the distribution method is DISTRIBUTE_BY_NONE.
However, soon this won't be the case when adding support for creating
single-shard distributed tables because their distribution method would
also be the same.

Now the internal method CreateCitusTable() doesn't make any assumptions
about table's replication model or such. Instead, it expects callers to
properly set all such metadata bits.

Even more, some of the parameters the old CreateDistributedTable() takes
 --such as the shard count-- were not meaningful for a reference table,
and would be the same as for new table type.

### CODE REVIEW

File: src/backend/distributed/commands/create_distributed_table.c [hunk 1]
```diff
@@ -951,17 +961,61 @@ EnsureRelationExists(Oid relationId)
...
  */
-void
-CreateDistributedTable(Oid relationId, char *distributionColumnName,
-					   char distributionMethod, int shardCount,
-					   bool shardCountIsStrict, char *colocateWithTableName)
+static void
+CreateCitusTable(Oid relationId, char *distributionColumnName,
+				 char distributionMethod, int shardCount,
+				 bool shardCountIsStrict, char *colocateWithTableName,
+				 char replicationModel)
```
Comment (marcocitus) [inline src/backend/distributed/commands/create_distributed_table.c:1018]: we should probably switch to a struct at some point, but not necessary now

File: src/backend/distributed/commands/create_distributed_table.c [hunk 2]
```diff
@@ -239,20 +225,48 @@ create_distributed_table(PG_FUNCTION_ARGS)
...
+		if (PG_ARGISNULL(2))
+		{
+			PG_RETURN_VOID();
+		}
 
-	char distributionMethod = LookupDistributionMethod(distributionMethodOid);
+		int shardCount = ShardCount;
+		if (!PG_ARGISNULL(4))
+		{
+			if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) != 0 &&
```
Comment (marcocitus) [inline src/backend/distributed/commands/create_distributed_table.c:239]: could be nice to use IsColocateWithNone and IsColocateWithDefault

File: src/backend/distributed/commands/create_distributed_table.c [hunk 3]
```diff
@@ -239,20 +225,48 @@ create_distributed_table(PG_FUNCTION_ARGS)
 
 	relation_close(relation, NoLock);
 
-	char *distributionColumnName = text_to_cstring(distributionColumnText);
-	Assert(distributionColumnName != NULL);
+	bool shardCountIsStrict = false;
+	if (distributionColumnText)
```
Comment (marcocitus) [inline src/backend/distributed/commands/create_distributed_table.c:229]: I don't understand why we have this condition. Maybe leave that out of this PR until we actually introduce logic related to distributionColumnText == NULL.

Comment (onurctirtir) [reply src/backend/distributed/commands/create_distributed_table.c]: moved these to https://github.com/citusdata/citus/pull/6745

File: src/backend/distributed/commands/create_distributed_table.c (hunk 2 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/create_distributed_table.c]: moved these to https://github.com/citusdata/citus/pull/6745 and addressed there

File: src/backend/distributed/commands/create_distributed_table.c [hunk 4]
```diff
@@ -1420,18 +1463,16 @@ DropFKeysRelationInvolvedWithTableType(Oid relationId, int tableTypeFlag)
...
 {
-	if (distributionMethod == DISTRIBUTE_BY_NONE)
-	{
-		return REPLICATION_MODEL_2PC;
-	}
-	else if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) != 0 &&
-			 !IsColocateWithNone(colocateWithTableName))
+	Assert(distributionMethod != DISTRIBUTE_BY_NONE);
+
+	if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) != 0 &&
```
Comment (onurctirtir) [inline src/backend/distributed/commands/create_distributed_table.c:1474]:
```suggestion
	if (!IsColocateWithDefault(colocateWithTableName) &&
```

File: src/backend/distributed/commands/create_distributed_table.c [hunk 5]
```diff
@@ -1420,18 +1463,16 @@ DropFKeysRelationInvolvedWithTableType(Oid relationId, int tableTypeFlag)
 
 
 /*
- * DecideReplicationModel function decides which replication model should be
- * used depending on given distribution configuration.
+ * DecideDistTableReplicationModel function decides which replication model should be
+ * used for given distributed table depending on given distribution configuration.
```
Comment (onurctirtir) [inline src/backend/distributed/commands/create_distributed_table.c:1467]:
```suggestion
 * used for a distributed table depending on given distribution configuration.
```

**Review by marcocitus — APPROVED** [review]

**Pull request merged**